### ISSUE TITLE
Unexpected warning when doc string invokes a macro which is defined within the same module

### ISSUE THREAD

**Issue by rami3l** [rust-lang/rust#133656]
This is a follow-up of https://github.com/rust-lang/rust/issues/124535#issuecomment-2466005846. 

TLDR: it looks like https://github.com/rust-lang/rust/issues/124535 hasn't been completely resolved after all...

### Issue Description
I've tried this code:

```rust
// lib.rs
macro_rules! pm_mods {
    ( $( $vis:vis $mod:ident; )+ ) => {
        $(
            $vis mod $mod;
            pub use self::$mod::$mod;
        )+
    }
}

pm_mods! {
    dnf;
}
```

```rust
// dnf.rs
#![doc = doc_self!()]

macro_rules! doc_self {
    () => {
        "The Dandified YUM."
    };
}
use doc_self; // <- Please note that the suggested fix has already been applied.

#[doc = doc_self!()]
pub fn dnf() {}
```

**I expected to see this happen**: ✅

**Instead, this happened**:

```console
> cargo check
    Checking repro_rust_124535 v0.1.0 (/path/censored)
warning: cannot find macro `doc_self` in this scope
 --> src/dnf.rs:1:10
  |
1 | #![doc = doc_self!()]
  |          ^^^^^^^^
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #124535 <https://github.com/rust-lang/rust/issues/124535>
  = help: import `macro_rules` with `use` to make it callable above its definition
  = note: `#[warn(out_of_scope_macro_calls)]` on by default

warning: `repro_rust_124535` (lib) generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.06s
```

Please refer to the minimal reproduction here: [rami3l/repro_rust_124535](https://github.com/rami3l/repro_rust_124535).

Also, it's worth noticing that after inlining the macro invocation in `lib.rs`, i.e.:

```rust
// lib.rs
mod dnf;
pub use self::dnf::dnf;
```

... the error seems gone.

### Meta

`rustc --version --verbose`:
```
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: aarch64-apple-darwin
release: 1.83.0
LLVM version: 19.1.1
```

**Comment by rami3l** [comment]
Additionally, I've come up with a similar single-file version that can be easily checked within the [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fb43f1d47e2f6c141e07a317c10fc2b1):

```rust
macro_rules! mods {
    ( $( $mod:ident; )+ ) => {
        $(
            mod $mod {
                #![doc = doc_self!()]

                macro_rules! doc_self {
                    () => { stringify!($mod) };
                }

                use doc_self; // <- Please note that the suggested fix has already been applied.
            }
        )+
    }
}

mods! {
    foo;
}
```

**Comment by jieyouxu** [comment]
cc @petrochenkov (in case you know if this is intended behavior or not)

**Comment by rami3l** [comment]
> cc [@petrochenkov](https://github.com/petrochenkov) (in case you know if this is intended behavior or not)

@jieyouxu Thanks a lot for the triage! To me this looks more like a false positive of the lint condition, given the fact that the `mods!{}` macro's expansion won't trigger the warning but its invocation will.

I'm not familiar with that part of the internals, but as long as the lint condition is not directly connected to the tree-visiting logic related to the upcoming breaking change, there is a possibility that this particular case will not break with that change in place, even if the warning says it will?

**Comment by rami3l** [comment]
Oops, with the latest version the same code has failed to compile :o
